### PR TITLE
[Structural] Defining initial state in shell elements

### LINKED ISSUE #9273: Implementing initial strains/stresses in shell elements
In order to simulate thermal expansion/contraction in structures we need to implement initial strains/stresses in shell and beam elements (we see that they are already implemented for solid elements.)
Starting with shell elements, it seems that the implementation should be quite straightforward, because the constitutive behaviour is defined using the `ConstitutiveLaw` class at the level of the integration points of the ply. Therefore a modification of `set_initial_state_process.cpp` for the case of shell elements to point to the ply integration points should suffice to achieve the desired behaviour.
In pseudo-code this should read something like this:
~~~c++
if (r_element.HasCrossSections()){
    loop for the sections in the element
        loop for the plies in the section
            loop for the integration points
                constitutive_law_vector[point_number]->SetInitialState(p_initial_state);

}
else{
    execute-existing-code
}
~~~
We would be glad if we get some support for this implementation, as we are Kratos newbies.
In particular we get stuck at how to point to the sections of the shell element, and how to point to the plies of the section.
For the first case it seems that a public method like
~~~c++
CrossSectionContainerType const& GetCrossSectionsOnIntegrationPoints() const
{
    return *mSections;
}
~~~
should be added to `base_shell_element.hpp` (and possibly to `element.h` as virtual method??)
And something similar for the plies:
~~~c++
PlyCollection const& GetPlyCollection() const
{
    return mStack;
}
~~~
We will be grateful for any help with this enhancement!

### PR DESCRIPTION
**Pull request by clazaro**
**📝 Description**
Including the possibility to define initial strains / stresses / deformation-gradient in shell elements.
Initial values corresponding to a 2D Plane Stress Law can be set using `set_initial_state_process.py`.
The current implementation allows to set a single initial value for the whole element. Setting different values for different plies in the shell cross section will require further changes.

Closes #9273

**🆕 Changelog**

Modified files:  `base_shell_element.h`, `base_shell_element.cpp`, `shell_cross_section.h`, `shell_cross_section.cpp` and `set_initial_state_process.cpp` to allow setting the intital state at the gauss points of the plies of the shell cross section.

Added test files to check the initial strain in a ShellQ4Thick element into `StructuralMechanicsApplication/tests/InitialStateElasticity/`
Added the corresponding class in `structural_mechanics_test_factory.py` and the test into `test_StructuralMechanicsApplication.py` nightSuite.

### CODE REVIEW

**Comment by clazaro** [comment]
I guess that this can be merged into master

File: applications/StructuralMechanicsApplication/custom_elements/base_shell_element.cpp
Comment (loumalouomega) [inline applications/StructuralMechanicsApplication/custom_elements/base_shell_element.cpp:519]: I would recommend to resize and the assign, instead of reserve + push_back

File: applications/StructuralMechanicsApplication/custom_utilities/shell_cross_section.cpp
Comment (loumalouomega) [inline applications/StructuralMechanicsApplication/custom_utilities/shell_cross_section.cpp:1310]: I woul assign size, it is known a a priori

**Review by loumalouomega — CHANGES_REQUESTED** [review]
Minor detail about vector sizes

File: applications/StructuralMechanicsApplication/custom_elements/base_shell_element.cpp
Comment (philbucher) [reply applications/StructuralMechanicsApplication/custom_elements/base_shell_element.cpp:519]: Why?

File: applications/StructuralMechanicsApplication/custom_utilities/shell_cross_section.cpp
Comment (philbucher) [reply applications/StructuralMechanicsApplication/custom_utilities/shell_cross_section.cpp:1310]: not directly like in the solids, as the section can have multiple layers
Hence I think it is not worth the effort of looping twice, esp since this fct is not called in performance critical parts of the code

File: applications/StructuralMechanicsApplication/custom_elements/base_shell_element.cpp
Comment (philbucher) [reply applications/StructuralMechanicsApplication/custom_elements/base_shell_element.cpp:519]: ah I remembered why I did it in this way, otherwise I would have to also store some indices
Since it is no difference performance wise I would leave it

File: applications/StructuralMechanicsApplication/custom_utilities/shell_cross_section.cpp
Comment (philbucher) [reply applications/StructuralMechanicsApplication/custom_utilities/shell_cross_section.cpp:1310]: I mean I could do it as above, but I really think it is not worth the effort ...

Comment (loumalouomega) [reply applications/StructuralMechanicsApplication/custom_utilities/shell_cross_section.cpp:1310]: Removing my block

**Review by loumalouomega — DISMISSED** [review]

File: applications/StructuralMechanicsApplication/custom_elements/base_shell_element.cpp
Comment (loumalouomega) [inline applications/StructuralMechanicsApplication/custom_elements/base_shell_element.cpp:517]:
```suggestion
        for (auto& p_sec : mSections) {
```

File: kratos/processes/set_initial_state_process.cpp
Comment (AlejandroCornejo) [inline kratos/processes/set_initial_state_process.cpp:155]: good point

**Review by AlejandroCornejo — APPROVED** [review]
as far as I am concerned, looks fine!

**Comment by AlejandroCornejo** [comment]
please wait for the confirmation of the rest before merging :)

**Review by philbucher — APPROVED** [review]

**Comment by clazaro** [comment]
Hi everyone. Should I wait for the technical committee review before merging?

**Comment by loumalouomega** [comment]
> Hi everyone. Should I wait for the technical committee review before merging?

As long as this does not affect to the core if the reviewers of the app agree, you should feel free to merge

**Comment by philbucher** [comment]
since the author code in the core that is affected by this PR already agreed I think you can go ahead
@rubenzorrilla any objections?

**Comment by clazaro** [comment]
Hi @rubenzorrilla, can I go ahead merging this PR?

**Pull request merged**
